lib/validate.js

```diff
@@ -0,0 +1,29 @@
+module.exports = validate
+
+var Promise = require('lie')
+
+function validate (state, doc) {
+  var self = this
+  return new Promise(function (resolve, reject) {
+    var result
+
+    try {
+      result = state.validate.call(self, doc)
+    } catch (error) {
+      reject(error)
+    }
+
+    if (result && result.then) {
+      return result.then(resolve, reject)
+    }
+
+    if (!result) {
+      var error = new Error('Document is not valid')
+
+      error.name = 'ValidationError'
+      reject(error)
+    }
+
+    resolve(doc)
```

Comment:
sorry I don’t quite understand? 

The validate method can, but does not need to be asynchronous. If it throws or returns a rejected promise, the validation will fail, otherwise it will succeed.

wether or not the validate method in your app sends requests to the server (is that what you mean with CRUD methods), is totally up to you